Fix four issues in archive.extracted state

terminalmage opened the pull request:
1. The trim_output argument was ignored for archives extracted using tarfile.

2. ``file.get_source_sum`` would fail if ``source`` is a list.

3. The checksum was being updated before we checked to see if it matched, effectively keeping us from detecting changes to the hash.

4. When ``source_hash_update`` is ``True``, and the archive is extracted, and files are later removed from the extraction dir, the state would not re-extract the archive unless the ``source_hash`` had changed (#39868).

Resolves #39868.

terminalmage committed:
Fix three issues in archive.extracted state

1. The trim_output argument was ignored for archives extracted using
   tarfile.

2. file.get_source_sum would fail if "source" is a list

3. The checksum was being updated before we checked to see if it
   matched, effectively keeping us from detecting changes to the hash.

4. When source_hash_update is True, and the archive is extracted, and
   files are later removed from the extraction dir, the state would not
   re-extract the archive unless the source_hash had changed.

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @nasenbaer13, @isbm and @magnusboman to be potential reviewers.

terminalmage committed:
Check hash of cached source against source_hash before downloading archive

This will force a re-download if the hash specified by source_hash if it
doesn't match the hash of the cached file, but if the hash matches, it
will skip the download. This will permit keep=True to prevent repeated
downloads of the source file, if and only if the source_hash also
remains the same.

isbm commented on salt/states/archive.py:
```diff
@@ -823,17 +834,20 @@ def extracted(name,
+                    'Invalid value for trim_output, must be True/False or an '
+                    'integer'
+                )
+                return ret
```
@terminalmage I'm just thinking maybe it is easier to comprehend by first discarding `bool`?
```python
trim_output = 100 if type(trim_output) == bool and trim_output else False
if trim_output and type(trim_output) != bool:
    try:
        trim_output = int(trim_output)
    except TypeError:
        ...
```

isbm reviewed (approved):
Seems OK (I wrote one comment first, but then deleted, since I was wrong). You love newlines and `\` though... :-)

terminalmage committed:
Just get a hash for the source archive

file.get_source_sum doesn't actually return a hash, it just
disambiguates a passed hash from a hash extracted from a hash file.

terminalmage commented:
What can I say, I like readability :man_shrugging:

cachedout commented:
@terminalmage Still pending changes here?

terminalmage commented:
@cachedout yes, I'm working with a community member to test updates to this state, the conversation is not taking place in this PR comment thread. I'll update the label when it's ready to merge.

terminalmage committed:
Remove unnecessary versionadded lines

There's already one at the top of the function's docstring.

terminalmage committed:
Moar fixes for source_hash_update

This commit adds additional fixes make source_hash_update work as
expected.

terminalmage commented:
I've pushed some changes that should have this ready for merging, but I'm waiting for testing/signoff from @jerrykan (see https://github.com/saltstack/salt/issues/40484#issuecomment-292661994).

terminalmage committed:
Fix mocking in unit tests

terminalmage commented:
Some failed unit tests due to the introduction of the ``hash_type`` option in the code (needed to be mocked in the tests).

terminalmage commented:
Should be fixed with my latest push.

Ch3LL commented:
@terminalmage is this ready for review and merge or are we still waiting on user?

terminalmage commented:
See #40484 for conversation with the community member I've been working with. It looks like I have one lingering issue to resolve in this PR before it can be merged.

terminalmage committed:
archive.extracted: also cleanup fileclient's cached location

When skip_verify is True, salt doesn't refetch a remote file it already
has in its cache. So, if "keep" is set to False, we need to remove it as
well as the location from file.managed.

terminalmage commented:
OK, this should be good to merge pending test results.
